components/omega/doc/design/Config.md

```diff
@@ -0,0 +1,410 @@
+models will have dependencies on other configuration choices. All
+configuration variables must therefore be accessible to all other
+modules/classes.
+
+### 2.5 Requirement: Support of data types
+
+Model configuration must support parameters in the standard data types,
+including logical (boolean), integer, float/double, and string data types.
+Limited support for vectors or arrays of the above types may be desirable.
+
+### 2.6 Requirement: Efficiency and parallelism
+
+Configuration should never consume significant resources, in either run
+time or storage. For parallel execution, the configuration
+will need to be replicated across MPI ranks, though in many cases, the
+variables will be manipulated into a different form (eg string variables
+describing choices will be converted to enums). This may happen before
+or after replication.
+
+### 2.8 Desired: Single configuration input
+
+For simplicity, we desire a single file for model configuration where users
+can define all model configuration details and provenance can be maintained
+more easily. If multiple files are required, there should be obvious links
+or includes in the main configuration file to make the dependency clear.
+
+### 2.9 Required: Hierarchy or grouping
+
+For both readability and for easier encapsulation by modules/classes, the 
+configuration should be organized in a logical way to make it clear where
+specific configuration variables are to be used. Parameters primarily or
+specifically associated with a module or class should be in a grouping
+of those parameters with the configuration.
+
+### 2.10 Requirement: Language support
+
+Because the configuration will also be used for provenance, it is likely
+that the configuration input file will need to be read by other languages
+outside of the OMEGA C++ model. This requires an ability to parse a 
+configuration input file from other common languages (eg python).
+
+### 2.11 Requirement: Optional or missing values
+
+If a configuration variable is missing from the configuration input file,
+there must be an option to either supply a default value or throw an error
+and exit, depending on the user choice. If a default value is supplied,
+then the default value must be added to the configuration so that it is
+included in the output and become part of the provenance.
+
+### 2.12 Requirement: Extra values
+
+If extra or unexpected values are encountered, they will be ignored.
+
+### 2.13 Desired: Automated generation of default input and error checking
+
+While the source code defines the configuration variables, it would
+desirable to have a means to extract from the source code what the 
+code is expecting into a default input config file. This would also
+enable some external error checking for missing or extra entries.
+
+### 2.14 Desired: Acceptable values
+
+For users modifying an input configuration, it would be desirable
+to document the acceptable values or range of values that each
+variable can be assigned. 
+
```

Comment:
Minor consideration on the ordering of key-value maps in YAML configuration file. I am not sure if Omega configuration needs to preserve the order of key-value maps, but it may be good to note here as YAML seems generally not support the ordering.

Reply:
@grnydawn Thanks - Ordering won't be important here. The plan is to read full config and extract as needed in no particular order.